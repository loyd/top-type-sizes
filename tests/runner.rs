use structopt::StructOpt;

use top_type_sizes::*;

fn test(content: &'static str) {
    let refined_content = reader::read(content.as_bytes()).unwrap();

    let types = parser::parse(&refined_content).unwrap();
    insta::with_settings!({ description => "internal representation" }, {
        insta::assert_yaml_snapshot!("types", types);
    });

    snap_output(&types, &[]);
    snap_output(&types, &["-ws"]);
    snap_output(&types, &["-h8"]);
    snap_output(&types, &["-ws", "-h16"]);
}

fn snap_output(types: &[schema::Type], cmd: &[&str]) {
    let full_cmd = std::iter::once(&"top-type-sizes")
        .chain(cmd)
        .copied()
        .collect::<Vec<_>>();

    let options = options::Options::from_iter(&full_cmd);
    let snap_name = {
        let mut parts = cmd.to_vec();
        parts.insert(0, "output");
        parts.join("")
    };

    let types = transformer::transform(types.to_vec(), &options);
    let output = formatter::format(types, &options);

    insta::with_settings!({ description => full_cmd.join(" ") }, {
        insta::assert_snapshot!(snap_name, output);
    })
}

macro_rules! test_sample {
    ($name:ident) => {
        #[test]
        fn $name() {
            insta::with_settings!(
            {
                snapshot_path => concat!("snapshots/", stringify!($name)),
                prepend_module_to_snapshot => false,
                omit_expression => true,
            },
            {
                test(include_str!(concat!("samples/", stringify!($name), ".txt")))
            })
        }
    };
}

test_sample!(command);
test_sample!(control_flow);
test_sample!(alignment_enum);
test_sample!(tokio_udp);
test_sample!(several_types);
test_sample!(timex);
test_sample!(async_fn);
test_sample!(compiler_messages);

// TODO: add samples from rustc tests.
