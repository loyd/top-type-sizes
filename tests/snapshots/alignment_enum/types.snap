---
source: tests/runner.rs
description: internal representation
---
- name: "std::ptr::alignment::AlignmentEnum64"
  size: 8
  align: 8
  kind:
    Enum:
      discriminant_size: 8
      variants:
        - name: _Align1Shl0
          size: 0
        - name: _Align1Shl1
          size: 0
        - name: _Align1Shl2
          size: 0
        - name: _Align1Shl3
          size: 0
        - name: _Align1Shl4
          size: 0
        - name: _Align1Shl5
          size: 0
        - name: _Align1Shl6
          size: 0
        - name: _Align1Shl7
          size: 0
        - name: _Align1Shl8
          size: 0
        - name: _Align1Shl9
          size: 0
        - name: _Align1Shl10
          size: 0
        - name: _Align1Shl11
          size: 0
        - name: _Align1Shl12
          size: 0
        - name: _Align1Shl13
          size: 0
        - name: _Align1Shl14
          size: 0
        - name: _Align1Shl15
          size: 0
        - name: _Align1Shl16
          size: 0
        - name: _Align1Shl17
          size: 0
        - name: _Align1Shl18
          size: 0
        - name: _Align1Shl19
          size: 0
        - name: _Align1Shl20
          size: 0
        - name: _Align1Shl21
          size: 0
        - name: _Align1Shl22
          size: 0
        - name: _Align1Shl23
          size: 0
        - name: _Align1Shl24
          size: 0
        - name: _Align1Shl25
          size: 0
        - name: _Align1Shl26
          size: 0
        - name: _Align1Shl27
          size: 0
        - name: _Align1Shl28
          size: 0
        - name: _Align1Shl29
          size: 0
        - name: _Align1Shl30
          size: 0
        - name: _Align1Shl31
          size: 0
        - name: _Align1Shl32
          size: 0
        - name: _Align1Shl33
          size: 0
        - name: _Align1Shl34
          size: 0
        - name: _Align1Shl35
          size: 0
        - name: _Align1Shl36
          size: 0
        - name: _Align1Shl37
          size: 0
        - name: _Align1Shl38
          size: 0
        - name: _Align1Shl39
          size: 0
        - name: _Align1Shl40
          size: 0
        - name: _Align1Shl41
          size: 0
        - name: _Align1Shl42
          size: 0
        - name: _Align1Shl43
          size: 0
        - name: _Align1Shl44
          size: 0
        - name: _Align1Shl45
          size: 0
        - name: _Align1Shl46
          size: 0
        - name: _Align1Shl47
          size: 0
        - name: _Align1Shl48
          size: 0
        - name: _Align1Shl49
          size: 0
        - name: _Align1Shl50
          size: 0
        - name: _Align1Shl51
          size: 0
        - name: _Align1Shl52
          size: 0
        - name: _Align1Shl53
          size: 0
        - name: _Align1Shl54
          size: 0
        - name: _Align1Shl55
          size: 0
        - name: _Align1Shl56
          size: 0
        - name: _Align1Shl57
          size: 0
        - name: _Align1Shl58
          size: 0
        - name: _Align1Shl59
          size: 0
        - name: _Align1Shl60
          size: 0
        - name: _Align1Shl61
          size: 0
        - name: _Align1Shl62
          size: 0
        - name: _Align1Shl63
          size: 0
